### PR TITLE
Add testing

### PR DESCRIPTION
**Pull request by JCZuurmond**
This PR is a proposal for adding testing using pytest. I would like to try this first here. If we are happy with the set-up, I want to move this into a python package so that it can be used for all `dbt-spark` projects.

### CODE REVIEW

File: test-requirements.txt
Comment (JCZuurmond) [inline test-requirements.txt:1]: The idea is that we use `pytest` and `pytest-spark` for testing `soda-spark` projects. `pytest` is the commonly used library for python testing. And `pytest-spark` allows you to test your spark applications.

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:26]: I have opened an [issue](https://github.com/dbt-labs/dbt-spark/issues/272) in `dbt-spark` to add a `pyspark` connection to the package. Then it is not needed to do this anymore.

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:86]: This is what I consider the power of this set-up: write a pytest unit test to test your dbt macros!

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:20]: See the [issue](https://github.com/dbt-labs/dbt-spark/issues/272) in `dbt-spark` about replacing the `Connection` with a reference to the `SparkSession`

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:23]: `dbt` complained about not having an `active_user` (it was `None`). If possible, I would like to remove this.

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:45]: `dbt` is written as a command line tool. The entrypoint of dbt - the dbt cli - uses the parsed arguments heavily. I tried to find a balance between reusing what is in `dbt-core` with minimal dependencies: here we still use the `from_args` class method to create a `RuntimeConfig` - but with a smaller `Args` class.

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:47]: This stores the Spark adapter in a global environment variable

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:53]: here we get the adapter again

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:55]: As mentioned before, I would like to move this into `dbt-spark`

**Review by JCZuurmond — COMMENTED** [review]
@jtcohen6 : I would love to hear what you think about this set-up! 

The main goal is to find a way to test our macros more easily. In expect that this could be a lightweight set-up to allow `dbt-spark` users to define pytest test for their macros. 

There are some integration points with the `dbt-core` I would like to get your feedback on.

File: tests/test_macros.py
Comment (jtcohen6) [reply tests/test_macros.py:185]: Really interesting, responded in the issue

File: tests/test_macros.py
Comment (jtcohen6) [reply tests/test_macros.py]: Heard. For our part, we know that:
- `RuntimeConfig` is a messy grab-bag
- We need to decouple the CLI / arg parsing logic from the actual entrypoints to the `dbt-core` application: https://github.com/dbt-labs/dbt-core/issues/4179 and https://github.com/dbt-labs/dbt-core/issues/4475

File: tests/test_macros.py
Comment (jtcohen6) [reply tests/test_macros.py]: This is really neat!!!

We've long wanted a way to write unit tests for dbt macros. This could open up better testing for:
- adapter plugin development (write + test each macro as you go)
- package development (such as this package)
- custom macro/materialization development in users' own projects

**Review by jtcohen6 — COMMENTED** [review]
@JCZuurmond Very neat, thanks for sketching this out! I remember your issue last year that encouraged us to lean much more into pytest built-ins (https://github.com/dbt-labs/dbt-core/issues/3193), reinventing fewer wheels and making it easier for external contributors (especially ones with some python know-how) to jump in.

As you mention, a lot of the initial code in this PR is easier setup/mocking for a dbt-spark connection in python. Let's continue that conversation over in https://github.com/dbt-labs/dbt-spark/issues/272. While it's obviously relevant to this PR/package, I sense it would be a much lighter lift for other adapters/databases.

So, the best bits are right at the end: I'm particularly taken by how you've used `@pytest.fixture` to mock macro inputs and outputs. This feels incredibly compelling to me. @gshank @emmyoop I'd be especially interested in getting your eyes and thoughts on this approach, as we think about our next-generation approach to testing dbt (with modularity and extensibilty top of mind).

**Comment by JCZuurmond** [comment]
As just said in https://github.com/dbt-labs/dbt-spark/issues/272, I will have a first go at a package for unit testing `dbt-spark` logic at the end of the month.

I would like some guidance on where to integrate with `dbt-core`. Let's discuss those in the comments above.

File: test-requirements.txt
Comment (JCZuurmond) [inline test-requirements.txt:3]: I moved all the set-up logic into a [package](https://pypi.org/project/pytest-dbt-core/)

**Review by JCZuurmond — COMMENTED** [review]
@jtcohen6 : Could you have another look. I have created a [package](https://pypi.org/project/pytest-dbt-core/) ([Github repo](https://github.com/godatadriven/pytest-dbt-core)) that contains the set-up logic for testing. It is written as a [Pytest plugin](https://docs.pytest.org/en/6.2.x/writing_plugins.html).

I think it would be best to add the Spark session stuff to the `dbt-spark` repo. This makes the `pytest-dbt-core` warehouse independent. I'll continue the discussion about this in [this issue](https://github.com/dbt-labs/dbt-spark/issues/272).

This PR is not ready to be merged until that issue is resolved.

**Comment by jtcohen6** [comment]
@JCZuurmond This is really neat!! Your work has sparked and anticipated many conversations internally, as we're working to refactor our own testing framework.

If you're interested, check out https://github.com/dbt-labs/dbt-core/pull/4691. We're just exiting the exploratory phase, and experimenting toward a set of concrete proposals. I'd love to hear your feedback over there, and (if you're up for it) find ways to include you in our ongoing work.

**Comment by JCZuurmond** [comment]
@jtcohen6 : Curious to get feedback again! Do you think this way of testing benefits the project? Will you merge this PR?

**Review by jtcohen6 — COMMENTED** [review]
@JCZuurmond I don't have any real opposition to including this sort of testing, as a neat experiment of what's possible for macro unit testing when you have a "database" that can run in-memory. Some neat possibilities for SQLite / DuckDB, perhaps? (cc @dbeatty10)

In practice, this package isn't one I'm able to regularly maintain, so I can't promise taking this any further than where you've brought it to date. I am interested in finding other folks who can help out — at dbt Labs, or potentially in the wider community.

File: profiles.yml
Comment (JCZuurmond) [reply profiles.yml:8]: Ok, I see this requirement comes from `dbt-spark`, not `dbt-core`. Do you have an example of a field that is conditionally typed on the connection type? I could make new `Credentials` data class, but don't understand how it will be used. The `Plugin` [has a reference](https://github.com/dbt-labs/dbt-spark/blob/5ff1c42560fbf4265d9cc36eb18f18fff0f91877/dbt/adapters/spark/__init__.py#L11) to the `SparkCredentials only.

File: tests/test_macros.py
Comment (JCZuurmond) [reply tests/test_macros.py:15]: Cool! I added [an example to the docs of `pytest-dbt-core](https://pytest-dbt-core.readthedocs.io/en/latest/) about renaming macros. That uses `pytest.mark.parameterize` to loop over different test scenarios. That is the power of this set-up to me: use of `pytest` features.

**Comment by JCZuurmond** [comment]
Thanks for reviewing the PR, @jtcohen6. I have **not** tested the library against other adapters than Spark. The lib is written adapter agnostic, but I expect it is easiest to use on in-memory databases. 

I am keen to include others on the project or to move parts of it to another project. If you could help with that, that would be great. I have ideas about what I want to change further, however, I prefer to include others first. To hear what others think and to allow the library to grow to be community based.

File: profiles.yml
Comment (JCZuurmond) [reply profiles.yml]: And do you want to wait with making this field optional in `dbt-spark`? Or are you ok with merging this PR and cleaning this up later?

Comment (jtcohen6) [reply profiles.yml]: No need for this to block the PR! Just wanted to record the clarification

**Review by jtcohen6 — COMMENTED** [review]
@JCZuurmond Last ask before we merge this: Could you update the README with a quick overview of this test? Plus, some code annotation in `tests/test_macros.py`?

File: tests/test_macros.py
Comment (JCZuurmond) [inline tests/test_macros.py:9]: @jtcohen6 : I moved this to a fixture so that the table is always deleted after a single test. This makes the test more clean and independent.

**Comment by JCZuurmond** [comment]
@jtcohen6 : Could you have another look?

**Comment by JCZuurmond** [comment]
The CI fails due to `tests/functional/test_utils.py` expecting a `dbt.tests.adapter` module.

File: test-requirements.txt
Comment (jtcohen6) [inline test-requirements.txt:2]: This is neat!

File: .github/workflows/workflow.yml
Comment (jtcohen6) [inline .github/workflows/workflow.yml:29]: I see - this tries to run `tests/functional/test_utils.py` as well (added in https://github.com/dbt-labs/spark-utils/pull/25), which has other requirements (specified in `dev-requirements.txt`).

Options:
1. **Unify:** Add `python -m pip install -r dev-requirements.txt` into the "Install dependencies" step above—or better yet, unify `dev-requirements` + `test-requirements` into a single file—and try running the tests with `--profile session`. I imagine that a few may fail (as in `dbt-spark`), but the rest should succeed, and we can mark the failing ones to skip on the "session" profile.
2. **Separate:** Treat these as two totally independent testing frameworks, which is really what they are. Create `tests/functional` for the tests using the `dbt-core` functional-testing framework, and `tests/unit` for the tests using the `pytest-dbt-core` unit-testing framework. Have clearly separate READMEs and files used by each: separate requirements, separate profile setup, all that jazz.

What do you think? I lean toward option 2, "separate" — and then `tests/unit` could be a clear, self-contained, and real-world demonstration of the potential of `pytest-dbt-core`

File: tests/test_macros.py
Comment (jtcohen6) [reply tests/test_macros.py:9]: I like it!

File: .github/workflows/workflow.yml
Comment (dbeatty10) [reply .github/workflows/workflow.yml:29]: +1 to option 2, "separate" (especially different folders `tests/unit` vs. `tests/functional`).

Separate folders won't undermine the functionality at all, and it will make the delineations between different frameworks more clear.

Side note: I don't see a problem with them sharing a single `dev-requirements.txt` though.

Comment (JCZuurmond) [reply .github/workflows/workflow.yml]: I moved the `test_macros.py` in a `unit` subfolder. And, I merged the `test-requirements.txt` into the `dev-requirements.txt`. I think it is confusing to have a `test-requirements.txt` that does not contain all dependencies for the functional tests. 

I had to move the `conftest.py` into the `functional` subdirectory because it was interfering with the unit tests.

**Review by jtcohen6 — APPROVED** [review]
Huzzah!! 🎉

@JCZuurmond Thanks for all your work to get this over the last mile. Excited to have this package as a demo of different approaches to testing complex macros.

**Pull request merged**
